Allow recurrence

wish2023 opened the pull request:
Implementations:
1. Weekly recurrence of tasks is possible.
2. User doesn't need to input date.

Todo:
Refactor relevant code into time class


Bugs:
1. Ui does not know the year when user types without the year, so we should consider getting the recurrence date in the AddCommand class itself, or we may consider to let the UI print only the month and day of the task added.

2. Ui needs to print different message when recurrence is day of the week. Eg: It says "It is set to happen tuesday" which is incorrect as it happens weekly.

Commit:
Add relevant libraries

Commit:
Create makDeepCopyTasks() and getDateAt() for debugging purposes

Commit:
Declare constants

Commit:
Add methods to remove date from recurrenceDates

Commit:
Change recurrenceDates to Arraylist to allow weekly repitition

Commit:
Rename getDate to getDates

Commit:
Allow forecast filter to filter through recurrence arraylist

Commit:
Allow weekly repetition of task

Commit:
Allow user to input date without year

Commit:
Allow TimeTable to display tasks using recurrenceDate Arraylist

Commit:
Follow code quality

Commit:
Merge branch 'master' into allow-recurrence

# Conflicts:
#	src/main/java/athena/task/Task.java
#	src/main/java/athena/task/Time.java

Kafcis commented on src/main/java/athena/task/Task.java:
```diff
@@ -71,22 +75,89 @@ public Task(String name, String startTime, String duration, String deadline,
+    private void setRecurrence(String recurrence) {
         //TODO: refactor this into Time class
 
         if (recurrence.toUpperCase().equals(Recurrence.TODAY.toString())) {
```
This if else block can be replaced with a switch block to improve readability

daniellimws replied:
and I think it is good enough to use `"TODAY"` instead of `Recurrence.TODAY.toString()`

Kafcis commented on src/main/java/athena/task/Task.java:
```diff
@@ -71,22 +75,89 @@ public Task(String name, String startTime, String duration, String deadline,
     }
 
+    private void addDates(LocalDate startDate) {
+        for (int i = 0; i < 10; i++) { // Max number of weeks now is 10 to prevent infinite recurrence
```
This is a good idea

Kafcis commented on src/main/java/athena/task/Task.java:
```diff
@@ -71,22 +75,89 @@ public Task(String name, String startTime, String duration, String deadline,
+
+    private LocalDate getFirstDay(DayOfWeek dayOfWeek) {
+        LocalDate startDate = LocalDate.now();
+        for (int i = 1; i < 7; i++) {
```
small nitpick, i can start from 0 then go to i<6, functionality wise is the same but might be more readable for others

Kafcis commented on src/main/java/athena/task/Task.java:
```diff
@@ -71,22 +75,89 @@ public Task(String name, String startTime, String duration, String deadline,
+        return LocalDate.parse(date, formatter);
+    }
+
+    public Task getClone() {
+        Task copy = new Task(name, startTime, duration, deadline, recurrence,
+                importance, notes, number, isFlexible);
```
not sure if we should support having tasks with the same task number, as long as the cloned tasks are not saved should be ok.

daniellimws replied:
I think it is better to override the `clone` method that is in `Object`.

```
    @Override
    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }
```

Kafcis commented on src/main/java/athena/task/Task.java:
```diff
@@ -110,16 +181,11 @@ public void edit(String name, String startTime, String duration,
-            }
-        }
-        assert !this.recurrenceDate.equals(null);
+        this.recurrence = recurrence;
+        resetRecurrence();
+        setRecurrence(recurrence);
```
Nice cleanup here, eventually should be moved to Time class

Kafcis reviewed (changes requested):
Looks good, the arraylist for the recurrence dates are going to be very convienent

daniellimws commented on src/main/java/athena/task/Task.java:
```diff
@@ -14,14 +16,16 @@
 public class Task {
     public static final String YES = "Y";
     public static final String NO = "N";
+    public static final int DATE_TIME_FORMAT = 5;
```
What does 5 refer to?

daniellimws replied:
Remember to remove this because it's no longer used.

daniellimws commented on src/main/java/athena/TaskList.java:
```diff
@@ -135,6 +139,10 @@ public String getTaskDescription(int taskNumber) throws TaskNotFoundException {
         return task.toString();
     }
 
+    public ArrayList<LocalDate> getDateAt(int i) {
+        return tasks.get(i).getDates();
+    }
```
`getDatesAt` instead? `Date` in plural.

Also, should we be using `getTaskFromNumber` instead?

wish2023 replied:
I just implemented this for debugging purposes I have removed it now

daniellimws commented on src/main/java/athena/task/Task.java:
```diff
@@ -71,22 +75,89 @@ public Task(String name, String startTime, String duration, String deadline,
+        }
+    }
+
+    private LocalDate getFirstDay(DayOfWeek dayOfWeek) {
```
`getFirstDay` is not so clear. Perhaps `getFirstDateMatchingDay`? Or something more clear?

daniellimws commented on src/main/java/athena/task/Task.java:
```diff
@@ -129,6 +195,13 @@ public void edit(String name, String startTime, String duration,
+            recurrenceDates.remove(FIRST_INDEX);
+        }
+    }
+
```
I think you can just do `recurrenceDates.clear()`

daniellimws commented on src/main/java/athena/task/Task.java:
```diff
@@ -213,8 +286,17 @@ public String getRecurrence() {
+     *
+     * @param date Date to delete
+     */
+    public void remove(LocalDate date) {
```
`removeDate` instead?

daniellimws commented on src/main/java/athena/task/Time.java:
```diff
@@ -37,6 +38,16 @@ public LocalDate getRecurrenceDate() {
 
+    private void setRecurrenceDate(String recurrence) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
+        if (recurrence.length() == DATE_TIME_FORMAT) {
```
`"dd-MM".length()` instead? `DATE_TIME_FORMAT` sounds quite confusing.

daniellimws commented on src/main/java/athena/task/Time.java:
```diff
@@ -37,6 +38,16 @@ public LocalDate getRecurrenceDate() {
+    private void setRecurrenceDate(String recurrence) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
+        if (recurrence.length() == DATE_TIME_FORMAT) {
+            String year = Integer.toString(LocalDate.now().getYear());
```
I suppose we should use the next year if the date has already passed. For example 1 Jan should be in year 2021.

daniellimws commented on src/main/java/athena/task/taskfilter/ImportanceFilter.java:
```diff
@@ -3,6 +3,8 @@
 import athena.Importance;
 import athena.task.Task;
 
+import java.time.LocalDate;
```
Unused import?

daniellimws commented on src/main/java/athena/task/taskfilter/FlexibleTimeFilter.java:
```diff
@@ -20,7 +20,7 @@ public FlexibleTimeFilter(Boolean isFlexible) {
     @Override
     public boolean isTaskIncluded(Task task) {
-        if (task.getDate() == null) {
+        if (task.getDates() == null) {
```
I think can use `task.isFlexible()` here, which has clearer intention than `task.getDates() == null`.

daniellimws commented on src/main/java/athena/task/taskfilter/ForecastFilter.java:
```diff
@@ -25,24 +26,45 @@ public ForecastFilter(Forecast forecast) {
+        return taskDate.get(woy);
+    }
+
+    private boolean isInstanceIncluded(LocalDate taskDate) {
```
Can this be `isDateIncluded` instead?

daniellimws commented on src/main/java/athena/task/taskfilter/ForecastFilter.java:
```diff
@@ -25,24 +26,45 @@ public ForecastFilter(Forecast forecast) {
-    private static int getWeekNumber(LocalDate taskDate) {
-        TemporalField woy = WeekFields.of(Locale.getDefault()).weekOfWeekBasedYear();
-        return taskDate.get(woy);
+    public Task filterDates(Task task) {
```
I think `filterDates` is not very clear. How about something like `removeExcludedDates`?

daniellimws reviewed (changes requested):
Nice! Left some comments about names which I felt to be ambiguous.

Commit:
Remove redundant libraries and methods

Commit:
Rename
1. isInstanceIncluded to isDateIncluded()
2. filterDates() to removeExcludedDates

Commit:
Get recurrence using switch case statements

daniellimws commented on src/main/java/athena/task/taskfilter/FlexibleTimeFilter.java:
```diff
@@ -20,12 +20,8 @@ public FlexibleTimeFilter(Boolean isFlexible) {
-            return isFlexible;
-        }
-        return !isFlexible;
+        return task.isFlexible();
```
The `isFlexible` attribute of this class is to determine whether to get flexible tasks or not flexible tasks. So, this needs to be `task.isFlexible() ^ isFlexible`. But anyways I think `isFlexible` is not very clear. How about something like `filterFlexibleTask` or `doesIncludeFlexibleTask`

daniellimws commented on src/main/java/athena/task/taskfilter/ForecastFilter.java:
```diff
@@ -25,24 +26,45 @@ public ForecastFilter(Forecast forecast) {
+    }
+
+    private boolean isDateIncluded(LocalDate taskDate) {
+        boolean isInstanceIncluded;
```
Change to `isDateIncluded` too?

daniellimws reviewed (changes requested):
`FlexibleTaskFilter` is currently not filtering tasks in the way as intended.

For the date related utility methods, I feel that `dateString` is a better name than `recurrence` since they are extracting information from a date. But anyways, we should refactor these things into the `Time` class later so we can skip them for now.

Also remember to add tests for:
- `ForecastFilter.removeExcludedDates`
- Check that recurring tasks are set to have the dates for the 10 upcoming weeks

Once these are done I think this is good to merge. Then we can work on refactoring these date related stuff into another class so that we don't clutter up the `Task` class.

Commit:
Create TaskTest

Commit:
Change setRecurrence() to public

Commit:
Update isTaskIncluded()

Commit:
Rename isInstanceIncluded to isDateIncluded

Commit:
Create tests for removeExcludedDates

Commit:
rename expectedTask to wrongTask for tests expected to be false

daniellimws commented on src/test/java/athena/task/TaskTest.java:
```diff
@@ -0,0 +1,51 @@
+     * Ensure repeating dates have been set correctly.
+     */
+    @Test
+    void testSetRecurrence_today_returnsTrue() {
```
I think it is unclear to say `today`, since `today` means only `TODAY` in our program? Something like`thisDayOfWeek` instead?

daniellimws commented on src/test/java/athena/task/TaskTest.java:
```diff
@@ -0,0 +1,51 @@
+     * Ensure repeating dates have not been set incorrectly.
+     */
+    @Test
+    void testSetRecurrence_today_returnsFalse() {
```
Same as above for `today`.

daniellimws commented on src/test/java/athena/task/taskfilter/ForecastFilterTest.java:
```diff
@@ -75,4 +78,77 @@ void testIsTaskIncluded_day_returnsFalse() {
+     * Check if relevant dates have been removed from task after filtering for a day.
+     */
+    @Test
+    void testremoveExcludedDates_day_returnsTrue() {
```
Slight nit: `testRemove` instead of `testremove`.
I also think `day` could be more descriptive, so that from the name we can tell what the test case is testing. How about `filterToday` or something that shows it is testing for filtering based on `TODAY`?

Also the method doesn't return `true` so this is not very precise. How about `returnsOnlyTodayDate`?

 Same goes for the methods below.

daniellimws commented on src/test/java/athena/task/taskfilter/ForecastFilterTest.java:
```diff
@@ -75,4 +78,77 @@ void testIsTaskIncluded_day_returnsFalse() {
+     */
+    @Test
+    void testremoveExcludedDates_week_returnsFalse() {
+        ForecastFilter forecastFilter = new ForecastFilter(Forecast.TODAY);
```
Is this supposed to be `Forecast.WEEK` instead, according to the comment?

daniellimws commented on src/test/java/athena/task/taskfilter/ForecastFilterTest.java:
```diff
@@ -75,4 +78,77 @@ void testIsTaskIncluded_day_returnsFalse() {
+        assertEquals(actualTask.equals(wrongTask), false);
+    }
+
+    private String getDateInString() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
+        return todayDate.plusDays(7).format(formatter);
+    }
```
Unused method? It doesn't seem to do exactly as the method name suggests too.

wish2023 replied:
Whoops forgot to delete

daniellimws reviewed (changes requested):
Looks good. Left some comments about the names. Once those are resolved I think it is good to merge.

Commit:
Rename test methods for clarity

Commit:
Merge branch 'master' into allow-recurrence

# Conflicts:
#	src/main/java/athena/timetable/Timetable.java
